src/Microsoft.AspNetCore.Blazor.Server/Circuits/RemoteRenderer.cs

```diff
@@ -12,10 +14,17 @@ namespace Microsoft.AspNetCore.Blazor.Browser.Rendering
 {
     internal class RemoteRenderer : Renderer
     {
+        // The purpose of the timeout is just to ensure server resources are released at some
+        // point if the client disconnects without sending back an ACK after a render
+        private const int TimeoutMilliseconds = 60 * 1000;
+
         private readonly int _id;
         private readonly IClientProxy _client;
         private readonly IJSRuntime _jsRuntime;
         private readonly RendererRegistry _rendererRegistry;
+        private readonly ConcurrentDictionary<long, AutoCancelTaskCompletionSource<object>> _pendingRenders
```

Comment:
One question that I have a this point - are we OK with out-of-order delivery? Or are we confident that SignalR will enforce ordering correctly. 

For instance `_client.SendAsync` could process an arbitrary number of batches before the browser sends any acks. 

Suppose that 1000000 renders get queued up on the server, and the client is taking the time to ack and process each one. Eventually I would expect some of the render operations to time out, and the order that those exceptions are delivered to the client is arbitrary.

This seems OK - I just wanted to mention it.

Reply:
AFAIK nothing depends on the ordering of batch completion notifications, so I think we are fine. Also in the success cases, the completion notifications are sure to be in order.